pasture-algorithms/src/normal_estimation.rs

```diff
@@ -0,0 +1,551 @@
+        centroid[3] = 1.0;
+    }
+}
+
+/// compute the covariance matrix for a given point cloud which is a measure of spread out the points are
+fn compute_covarianz_matrix<T: PointBuffer>(
+    point_cloud: &T,
+    covariance_matrix: &mut DMatrix<f64>,
+) -> usize {
+    let mut point_count = 0;
+    if point_cloud.is_empty() {
+        return 0;
+    }
+    // compute the centroid of the point cloud
+    let mut centroid = Vector4::<f64>::zeros();
+    compute_centroid(point_cloud, &mut centroid);
+
+    if is_dense(point_cloud) {
+        point_count = point_cloud.len();
+        let mut diff_mean = vec![0.0; 4];
+        for point in point_cloud.iter_attribute::<Vector3<f64>>(&POSITION_3D) {
+            // calculate difference from the centroid for each point
+            diff_mean[0] = point.x - centroid[0];
+            diff_mean[1] = point.y - centroid[1];
+            diff_mean[2] = point.z - centroid[2];
+
+            covariance_matrix[(1, 1)] += diff_mean[1] * diff_mean[1];
+            covariance_matrix[(1, 2)] += diff_mean[1] * diff_mean[2];
+            covariance_matrix[(2, 2)] += diff_mean[2] * diff_mean[2];
+
+            let diff_x = diff_mean[0];
+            diff_mean.iter_mut().for_each(|x| *x = *x * diff_x);
+
+            covariance_matrix[(0, 0)] += diff_mean[0];
+            covariance_matrix[(0, 1)] += diff_mean[1];
+            covariance_matrix[(0, 2)] += diff_mean[2];
+        }
+    } else {
+        // in this case we dont know the number of points in the point cloud that are finite
+        for point in point_cloud.iter_attribute::<Vector3<f64>>(&POSITION_3D) {
+            if !is_finite(&point) {
+                continue;
+            }
+            // only compute the covariance matrix for finite points
+            let mut diff_mean = vec![0.0; 4];
+            // calculate difference from the centroid for each point
+            diff_mean[0] = point.x - centroid[0];
+            diff_mean[1] = point.y - centroid[1];
+            diff_mean[2] = point.z - centroid[2];
+
+            covariance_matrix[(1, 1)] += diff_mean[1] * diff_mean[1];
+            covariance_matrix[(1, 2)] += diff_mean[1] * diff_mean[2];
+            covariance_matrix[(2, 2)] += diff_mean[2] * diff_mean[2];
+
+            let diff_x = diff_mean[0];
+            diff_mean.iter_mut().for_each(|x| *x = *x * diff_x);
+
+            covariance_matrix[(0, 0)] += diff_mean[0];
+            covariance_matrix[(0, 1)] += diff_mean[1];
+            covariance_matrix[(0, 2)] += diff_mean[2];
+            point_count = point_count + 1;
+        }
+    }
+
+    covariance_matrix[(1, 0)] = covariance_matrix[(0, 1)];
+    covariance_matrix[(2, 0)] = covariance_matrix[(0, 2)];
+    covariance_matrix[(2, 1)] = covariance_matrix[(1, 2)];
+
+    return point_count;
+}
+
+/// find the eigen value solution if the highest degree of the polynomial is 2
+fn solve_polynomial_quadratic(
+    coefficient_2: &f64,
```

Comment:
You don't have to pass `f64` by reference.
If you know that there are exactly 3 Eigenvalues, you can also return `Vector3<f64>`. In any case, the eigenvalues should be returned instead of passed in